File: 01-introduction-and-complexity/01-asymptotic-notation/javascript/asymptoticDemo.test.js
/* 01 漸進符號示範測試（Node.js）/ Tests for asymptotic demo (Node.js). */  // Bilingual test file header.
'use strict';  // Enable strict mode for consistent runtime behavior.

const test = require('node:test');  // Use Node's built-in test runner (no external dependencies).  
const assert = require('node:assert/strict');  // Use strict assertions for clearer failures.  

const {  // Import the demo functions under test from the local module.
  countConstantOps,  // O(1) counter under test.
  countLog2Ops,  // O(log n) counter under test.
  countLinearOps,  // O(n) counter under test.
  countNLog2NOps,  // O(n log n) counter under test.
  countQuadraticOps,  // O(n^2) counter under test.
} = require('./asymptoticDemo');  // Load the demo module from the same folder.

test('countConstantOps should be constant', () => {  // Verify the O(1) demo ignores input size.
  assert.equal(countConstantOps(0), countConstantOps(10));  // Same output for different inputs.
  assert.equal(countConstantOps(1), 3);  // This demo uses exactly 3 simulated operations.
});  // End test case.

test('countLog2Ops should match halving count', () => {  // Verify the O(log n) demo returns floor(log2(n)).
  assert.equal(countLog2Ops(1), 0);  // No halving needed when n == 1.
  assert.equal(countLog2Ops(2), 1);  // 2 -> 1 uses one halving.
  assert.equal(countLog2Ops(8), 3);  // 8 -> 4 -> 2 -> 1 uses three halvings.
});  // End test case.

test('countLog2Ops should reject non-positive', () => {  // Verify invalid inputs throw for log-based demo.
  assert.throws(() => countLog2Ops(0), /n must be >= 1/);  // log2(0) is not accepted in this discrete demo.
  assert.throws(() => countLog2Ops(-1), /n must be >= 1/);  // Negative sizes are invalid.
});  // End test case.

test('countLinearOps should equal n', () => {  // Verify the O(n) demo returns exactly n operations.
  assert.equal(countLinearOps(0), 0);  // Zero-sized input performs zero operations.
  assert.equal(countLinearOps(5), 5);  // One operation per iteration for n iterations.
});  // End test case.

test('countQuadraticOps should equal n^2', () => {  // Verify the O(n^2) demo returns n squared operations.
  assert.equal(countQuadraticOps(0), 0);  // 0 * 0 iterations.
  assert.equal(countQuadraticOps(4), 16);  // 4 * 4 iterations.
});  // End test case.

test('countNLog2NOps should equal n * floor(log2(n)) for powers of two', () => {  // Verify the intended n log n growth model.
  assert.equal(countNLog2NOps(0), 0);  // Define 0 * log(0) as 0 work for this demo.
  assert.equal(countNLog2NOps(8), 24);  // 8 * log2(8) = 8 * 3 = 24 operations.
});  // End test case.

File: 01-introduction-and-complexity/01-asymptotic-notation/java/AsymptoticDemoTest.java
/** 01 漸進符號示範測試（Java）/ Tests for asymptotic demo (Java). */  // Bilingual file header for tests.
public final class AsymptoticDemoTest {  // Declare a tiny test runner without external libraries.
    private AsymptoticDemoTest() {  // Prevent instantiation because this class is a test harness.
        // No instances.  // Explain why the constructor is private.
    }  // Close the constructor block.

    private static void assertEquals(long expected, long actual, String message) {  // Assert helper for numeric equality.
        if (expected != actual) {  // Fail when the values differ.
            throw new AssertionError(message + " (expected=" + expected + ", actual=" + actual + ")");  // Report a clear mismatch message.
        }  // Close the assertion block.
    }  // End of assertEquals.

    private static void assertThrows(Class<? extends Throwable> expectedType, Runnable action, String message) {  // Assert helper for exception behavior.
        try {  // Execute the action and ensure it throws.
            action.run();  // Run the action that is expected to fail.
        } catch (Throwable ex) {  // Capture any thrown exception so we can verify its type.
            if (expectedType.isInstance(ex)) {  // Accept when the thrown type matches the expected type.
                return;  // Return early because the test passed.
            }  // Close the type-check block.
            throw new AssertionError(message + " (wrong exception type: " + ex.getClass().getName() + ")");  // Fail if the exception type is wrong.
        }  // End of catch block.
        throw new AssertionError(message + " (no exception was thrown)");  // Fail if the action did not throw at all.
    }  // End of assertThrows.

    private static void testConstantOpsIsConstant() {  // Verify that the O(1) counter is independent of n.
        assertEquals(AsymptoticDemo.countConstantOps(0), AsymptoticDemo.countConstantOps(10), "O(1) should be constant");  // Compare different inputs.
        assertEquals(3, AsymptoticDemo.countConstantOps(1), "This demo uses exactly 3 operations");  // Confirm the chosen constant.
    }  // End of testConstantOpsIsConstant.

    private static void testLog2OpsMatchesHalvingCount() {  // Verify that the halving loop counts floor(log2(n)).
        assertEquals(0, AsymptoticDemo.countLog2Ops(1), "log2 ops for n=1 should be 0");  // No halving for n=1.
        assertEquals(1, AsymptoticDemo.countLog2Ops(2), "log2 ops for n=2 should be 1");  // 2 -> 1.
        assertEquals(3, AsymptoticDemo.countLog2Ops(8), "log2 ops for n=8 should be 3");  // 8 -> 4 -> 2 -> 1.
    }  // End of testLog2OpsMatchesHalvingCount.

    private static void testLog2OpsRejectsNonPositive() {  // Verify that invalid inputs are rejected for the log demo.
        assertThrows(IllegalArgumentException.class, () -> AsymptoticDemo.countLog2Ops(0), "log2 ops should reject n=0");  // log2(0) is not accepted here.
        assertThrows(IllegalArgumentException.class, () -> AsymptoticDemo.countLog2Ops(-1), "log2 ops should reject negative n");  // Negative sizes are invalid.
    }  // End of testLog2OpsRejectsNonPositive.

    private static void testLinearOpsEqualsN() {  // Verify that the linear counter returns exactly n operations.
        assertEquals(0, AsymptoticDemo.countLinearOps(0), "linear ops for n=0 should be 0");  // 0 iterations.
        assertEquals(5, AsymptoticDemo.countLinearOps(5), "linear ops for n=5 should be 5");  // 5 iterations.
    }  // End of testLinearOpsEqualsN.

    private static void testQuadraticOpsEqualsNSquared() {  // Verify that the quadratic counter returns n^2 operations.
        assertEquals(0, AsymptoticDemo.countQuadraticOps(0), "quadratic ops for n=0 should be 0");  // 0 * 0.
        assertEquals(16, AsymptoticDemo.countQuadraticOps(4), "quadratic ops for n=4 should be 16");  // 4 * 4.
    }  // End of testQuadraticOpsEqualsNSquared.

    private static void testNLogNOpsEqualsNTimesLog2N() {  // Verify that the n log n counter matches n * floor(log2(n)) for powers of two.
        assertEquals(0, AsymptoticDemo.countNLog2NOps(0), "n log n ops for n=0 should be 0");  // Boundary case definition.
        assertEquals(24, AsymptoticDemo.countNLog2NOps(8), "n log n ops for n=8 should be 24");  // 8 * log2(8) = 8 * 3.
    }  // End of testNLogNOpsEqualsNTimesLog2N.

    public static void main(String[] args) {  // Execute all tests as a simple CLI program.
        testConstantOpsIsConstant();  // Run the O(1) test case.
        testLog2OpsMatchesHalvingCount();  // Run the O(log n) test case.
        testLog2OpsRejectsNonPositive();  // Run invalid-input tests for the log demo.
        testLinearOpsEqualsN();  // Run the O(n) test case.
        testQuadraticOpsEqualsNSquared();  // Run the O(n^2) test case.
        testNLogNOpsEqualsNTimesLog2N();  // Run the O(n log n) test case.
        System.out.println("All tests PASSED.");  // Print a success message when no assertions fail.
    }  // End of test runner main.
}  // End of AsymptoticDemoTest class.

